### PR TITLE
[REF] *: use hoot events in tour_utils 

### PR DESCRIPTION
**Pull request by pipu-odoo**
In this commit, events in
addons/web_tour/static/src/tour_service/tour_utils.js writed with native dispatchEvents has been refactored to use Hoot events. These events are much closer to the behavior of a user (e.g. instead of overwriting the value of an input, it simulates the encoding using keydowns).
The motivation of these changes is to make the tours more understandable and significantly less implicit. As a result, “automatic” behaviors have been removed.

Here below are the main modifications made in RunningTourActionHelper class : 
- The auto method has been removed.
- remove_text() has been replaced by clear(). 
- text() has been splitted in edit(), fill(), select(), selectByIndex(), selectByLabel() and editor(). 
- Default "Test" string has been removed. No text => No text (Then, note that `run edit` will just clear the input) 
- New events has been added like range(), press() and blur(). 

A New feature has been added in tour_compiler.js : 
You can now chain events in run attribute of a step with `&&` : 
e.g. : 
```
{
    extra_trigger: ".o_web_studio_sidebar .o_web_studio_properties.active",
    trigger: "input[name='string']",
    run: "edit new name && press Enter",
},
```

task~3820046

https://github.com/odoo/enterprise/pull/59337

### CODE REVIEW

File: addons/test_website/static/tests/tours/website_settings.js [hunk 1]
```diff
@@ -26,6 +26,7 @@ registry.category("web_tour.tours").add("website_settings_m2o_dirty", {
         {
             content: "open website switcher",
             trigger: "input[id^='website_id']",
+            run: "edit Test",
```
Comment (rdeodoo) [inline addons/test_website/static/tests/tours/website_settings.js:29]: Can be ignored if you pref but:
```suggestion
            run: `edit {websiteName}`,
```
or
```suggestion
            run: "edit Website Test",
```
Would imo make it more clear now that's it's explicit, I was wondering what was "Test"

File: addons/website/static/tests/tours/edit_link_popover.js [hunk 2]
```diff
@@ -47,7 +47,7 @@ wTourUtils.registerWebsitePreviewTour('edit_link_popover', {
     {
         content: "Type the link URL /contactus",
         trigger: '#o_link_dialog_url_input',
-        run: 'text /contactus'
+        run: "edit /contactus && blur we-title.o_short_title",
```
Comment (rdeodoo) [inline addons/website/static/tests/tours/edit_link_popover.js:50]: Is the blur really needed here and at line 66 + 106?
Won't it be automatically triggered when clicking on the footer or the save button in the next step?

File: addons/website/static/tests/tours/snippet_popup_display_on_click.js [hunk 3]
```diff
@@ -70,7 +70,7 @@ wTourUtils.registerWebsitePreviewTour("snippet_popup_display_on_click", {
     {
         content: "Add a link to the homepage in the URL input",
         trigger: "#o_link_dialog_url_input",
-        run: "text /"
+        run: "focus && press /",
```
Comment (rdeodoo) [inline addons/website/static/tests/tours/snippet_popup_display_on_click.js:73]: Why can't it be
```suggestion
        run: "edit /",
```

**Review by rdeodoo — CHANGES_REQUESTED** [review]
Nice :+1:  I am just commenting the website part

File: addons/website/tests/test_ui.py [hunk 4]
```diff
@@ -298,7 +298,7 @@ def test_08_website_style_custo(self):
         self.start_tour(self.env['website'].get_client_action_url('/'), 'website_style_edition', login='admin')
 
     def test_09_website_edit_link_popover(self):
-        self.start_tour('/@/', 'edit_link_popover', login='admin')
+        self.start_tour('/@/', 'edit_link_popover', login='admin', timeout=250, step_delay=100)
```
[the remark this answers is not in the document]
Comment (pipu-odoo) [reply addons/website/tests/test_ui.py:301]: When I run this tour in watch(debug) mode, I have no problems. But when I run it in headless mode, tour fails. 
the only way to successfully make it run is to modify these parameters. With the tour above (website_form_editor_tour), these are the only 2 tours that do not work with the modifications made other than modifying the step_delay.
I spent a full day trying to find a solution for these 2 rounds... this is the only solution I found without changing the tour definition.

File: addons/website/static/tests/tours/edit_link_popover.js [hunk 5]
```diff
@@ -145,11 +145,16 @@ wTourUtils.registerWebsitePreviewTour('edit_link_popover', {
...
+    },
+    {
+        content: "Ensure popover is closed",
+        trigger: ':iframe html:not(:has(.o_edit_menu_popover))', // popover should be closed
         run: function () {}, // it's a check
     },
     {
         content: "Click 'Home' link in footer",
-        trigger: ':iframe footer a[href="/"]',
+        trigger: `:iframe footer a[href="/"]`,
```
[the remark this answers is not in the document]
Comment (pipu-odoo) [reply addons/website/static/tests/tours/edit_link_popover.js:157]: Yes, I should remove this step ... not useful and as you say, this is not the same behavior as before.

File: addons/website/static/tests/tours/snippet_popup_display_on_click.js (hunk 3 above)
Comment (pipu-odoo) [reply addons/website/static/tests/tours/snippet_popup_display_on_click.js:73]: Yes, it's the same :D ...

File: addons/website/static/tests/tours/edit_link_popover.js (hunk 2 above)
Comment (pipu-odoo) [reply addons/website/static/tests/tours/edit_link_popover.js:50]: You're right ! I'll remove all modifications in popover tour. There are leftovers of my tests to debug this tour...

File: addons/website/tests/test_ui.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (rdeodoo) [reply addons/website/tests/test_ui.py:301]: 1. Doesn't that mean there is an issue with the new tour engine?
2. isn't the step_delay only needed on one or two steps and not each steps of the tour?
3. Does the tour timeout really need to be increased that much, even with the step_delay? The tour doesn't seem to be taking more than 30 seconds on your runbot build I believe

File: addons/test_website/static/tests/tours/website_settings.js (hunk 1 above)
Comment (pipu-odoo) [reply addons/test_website/static/tests/tours/website_settings.js:29]: Done

File: addons/website/tests/test_website_form_editor.py [hunk 6]
```diff
@@ -20,7 +20,7 @@ def setUpClass(cls):
         })
 
     def test_tour(self):
-        self.start_tour(self.env['website'].get_client_action_url('/'), 'website_form_editor_tour', login='admin', timeout=120)
+        self.start_tour(self.env['website'].get_client_action_url('/'), 'website_form_editor_tour', timeout=250, step_delay=100, login='admin')
```
[the remark this answers is not in the document]
Comment (pipu-odoo) [reply addons/website/tests/test_website_form_editor.py]: Thank you for your help :) ... Resolved !

**Review by rdeodoo — APPROVED** [review]
Approving the website part, thanks for the changes.
For tracking purpose, the 'create new' button tweak here is because it simply does not work in human/manual flow. Now that hoot is closer to real human flow, it highlight the bug which will be fixed later.

File: addons/web_tour/static/src/tour_service/tour_utils.js [hunk 7]
```diff
@@ -150,129 +149,145 @@ export const triggerPointerEvent = (el, type, canBubbleAndBeCanceled, additional
...
+        if (selector?.length) {
+            hoot.click(selector);
+        } else if (this.previousActiveElement !== this.currentActiveElement) {
+            hoot.click(this.previousActiveElement);
+        } else {
+            hoot.click("body");
+        }
+    }
+
+    clear = (selector) => {
```
[the remark this answers is not in the document]
Comment (pipu-odoo) [reply addons/web_tour/static/src/tour_service/tour_utils.js:167]: ok

**Pull request closed without merging**